src/gd_avif.c

```diff
@@ -0,0 +1,667 @@
+    from a gdIOCtx. So, here, we use our own helper function, createAvifIOfromCtx().
+
+    Otherwise, we create the image by calling AVIF library functions in order:
+    * avifDecoderCreate(), to create the decoder
+    * avifDecoderSetIO(), to tell libavif how to read from our data structure
+    * avifDecoderParse(), to parse the image
+    * avifDecoderNextImage(), to read the first image from the decoder
+    * avifRGBImageSetDefaults(), to create the avifRGBImage
+    * avifRGBImageAllocatePixels(), to allocate memory for the pixels
+    * avifImageYUVToRGB(), to convert YUV to RGB
+    
+    Finally, we create a new gd image and copy over the pixel data.
+  
+  Parameters:
+
+    ctx              - a gdIOCtx struct
+*/
+BGD_DECLARE(gdImagePtr) gdImageCreateFromAvifCtx (gdIOCtx * ctx)
+{
+  int x, y;
+  uint8_t *p8; 
+  uint16_t *p16;
+  gdImage *im = NULL;
+  avifResult result;
+  avifIO *io;
+  avifDecoder *decoder;
+  avifRGBImage rgb;
+
+  decoder = avifDecoderCreate();
+  io = createAvifIOFromCtx(ctx);
+
+  // memset(rgb, 0, sizeof(rgb));
+
+  avifDecoderSetIO(decoder, io);
+
+  result = avifDecoderParse(decoder);
+  if (isAvifError(result, "Could not parse image")) {
+    goto cleanup;
+  }
+
+  // Note again that, for an image sequence, we read only the first image, ignoring the rest.
+  result = avifDecoderNextImage(decoder);
+  if (isAvifError(result, "Could not decode image")) {
+    goto cleanup;
+  }
+
+  // Set up the avifRGBImage. Use default settings unless there's demand to override these.
+  avifRGBImageSetDefaults(&rgb, decoder->image);
+  avifRGBImageAllocatePixels(&rgb);
+
+  result = avifImageYUVToRGB(decoder->image, &rgb);
+  if (isAvifError(result, "gd-avif error: Conversion from YUV to RGB failed")) {
+    goto cleanup;
+  }
+
+  im = gdImageCreateTrueColor(decoder->image->width, decoder->image->height);
+	if (!im) {
+		gd_error("avif error: Could not create GD truecolor image");
+		goto cleanup;
+	}
+
+
+// Read the pixels from the AVIF image and copy them into the GD image.
+// Image depth can be 8, 10, 12, or 16. But if depth>8, pixels are 16-bit.
+
+  if (rgb.depth == 8) {
+    for (y = 0, p8 = rgb.pixels; y < decoder->image->height; y++) {
+      for (x = 0; x < decoder->image->width; x++) {
+        uint8_t r = *(p8++);
+        uint8_t g = *(p8++);
+        uint8_t b = *(p8++);
+        uint8_t a = *(p8++);
+        im->tpixels[y][x] = gdTrueColorAlpha(r, g, b, a);
+      }
+    }
+    
+  } else {
+    for (y = 0, p16 = (uint16_t *) rgb.pixels; y < decoder->image->height; y++) {
+      for (x = 0; x < decoder->image->width; x++) {
+        uint16_t r = *(p16++);
+        uint16_t g = *(p16++);
+        uint16_t b = *(p16++);
+        uint16_t a = *(p16++);
+        im->tpixels[y][x] = gdTrueColorAlpha(r, g, b, a);
+      }
+    }
+  }
+
+  /* do not use gdFree here, in case gdFree/alloc is mapped to something else than libc */
```

Comment:
I completely miss color profile handling here, ICC or NCLX boxes ...

Reply:
Reading your comment, I looked around the libgd codebase, and I didn't see explicit support for color profiles anywhere else - not even in PNGs, where I believe this is included in the spec.

Of course, libavif supports embedded ICC profiles explicitly.

But since it's not something gd deals with, I'm not sure if it would make sense to add it here. How do you think this should be handled?

Reply:
i don't think we want to hold up adding libavif support for color profiles if we aren't doing color profiles anywhere yet

Reply:
I would say this is a big problem. AVIF has been created to be able to handle HDR images e.g. P3 PQ or HG. It means we can only handle sRGB files.

So we have to check the NCLX box and if it isn't sRGB stop and if it has an ICC profile we have to abort too.

Here are test files:
https://github.com/AOMediaCodec/av1-avif/tree/master/testFiles/Netflix/avif

Reply:
i don't think we're disagreeing that we should have color profile support in general, but we are disagreeing that it needs to be implemented *before* merging libavif support.  feel free to comment on #136.

Reply:
Do you think we need to check the icc field when decoding an AVIF image - and abort if one is found? Or can we just leave this for now? I don't see a way that the NCLX status is stored explicitly.

If it helps, I believe the primary use case here will be for web images. Though @vapier would know more.

Reply:
I did not say you should add support for color profiles here, however if the AVIF file isn't using sRGB which is the web standard it doesn't make sense to convert to any other image format as the result will look just horrible on your screen.

So when coding check if the NCLX box says it is sRGB we fine. If it isn't sRGB you will end up with garbage (horribly looking image).

Reply:
ok - sounds like I should make sure the color profile is sRGB, and, if it's not, throw an error.

Reply:
I looked at this for a little while tonight. Trying decoding on an image that doesn't have ICC info, I saw this:

![Screen Shot 2021-02-10 at 8 42 55 PM](https://user-images.githubusercontent.com/29419498/107595325-54895780-6be2-11eb-92b9-5b2705c7f1ee.png)

I could certainly throw an error if values were set here that did *not* correspond to sRGB. If I got sRGB values, or I got the `UNSPECIFIED` values in the screenshot, I wouldn't throw an error.

To see how this struct looks for sRGB images, I looked at the test files you sent. As far as I can tell, the AVIF images were encoded without ICC:

`avifenc -c aom --speed 0 --jobs 12 --depth 8 --ignore-icc ...`

So I downloaded an HDR PNG and encoded that to AVIF, then ran it though my decoder without that `--ignore-icc` flag. Sure enough, I saw other values:

![SDR image](https://user-images.githubusercontent.com/29419498/107599669-4988f400-6bef-11eb-9f53-ed5bda100333.png)

Presumably, if I saw this, I'd throw an error.

I then downloaded an SDR PNG from the same repo and encoded to AVIF, and ran it through the decoder... but this time I got the same values as I did from the HDR PNG. Hmm.

Am I doing this correctly? Do you have an AVIF image that's got an sRGB ICC profile attached that you could send me for testing?

Reply:
If you convert a png to an avif and an ICC profile is attached to the png, you should just copy the icc blob over to the avif file. You don't know its content. If there is no ICC you should just assume it is sRGB and write a sRGB NCLX box. If you have an AVIF with an ICC profile, convert it and write the ICC profile to the image ...

An NCLX box holds the same information as an ICC profile. However we are not able to convert an NCLX box to an ICC profile or vice versa. At least right now. You would need littlecms2 for this.

If you decode an AVIF file and it has an NCLX box which is not sRGB we need to stop. It doesn't make sense to convert the image as the resulting file can be correctly displayed by any software.

Reply:
I’m just a user, so don’t know much, please treat this comment with a truckload of salt. As noted, the library is currently fully colour blind. Regarding stopping processing on anything non-sRGB, this is quite drastic and would make it unpredictable in any automated scenario (some files would just bomb). AVIF aren’t any different to other files whose profiles library currently is blind to, if maybe for the fact that AVIF is more likely to have exotic, wide gamut data. The further source data is from sRGB, the more insane is the output, but, again, it also applies to any other format (eg. a PNG with ProPhotoRGB converted to a PNG).

>If you convert a png to an avif and an ICC profile is attached to the png, you should just copy the icc blob over to the avif file.

This is scenario no. 1 [from here](https://github.com/libgd/libgd/issues/136#issuecomment-318921312). While not exactly correct, it’s less incorrect than current situation (dropping the profile). Would be great if this behaviour could apply to all images, not just AVIF. “Great”, that is, until the library becomes fully colour-aware at some point in the future.

Reply:
"I'm just a user" - that means you're a vital part of this conversation! Thanks for putting this here.

I've implemented @cryptomilk's suggestion to detect non-sRGB images and throw an error, but I'm honestly more comfortable with what you suggest. I feel that someone who decides to convert a non-sRGB image shouldn't be surprised if the result isn't great.

I see that `gd_errors.h` allows for warnings, not just fatal errors - and that files like `gd_jpeg.c` and `gd_png.c` use warnings. Perhaps that a warning would be a good compromise? Even though I don't think warnings would be visible to people who use GD via, say, the PHP wrapper.

Reply:
Changed this to emit a warning.